File: BackTracking/wordBreak.py
from itertools import groupby
class Solution:
    def wordBreak(self, n, d, s):
        # code here
        res = []

        def util(pos, ip, op):
            if len(op) > 0 and op[-1] not in d:
                return
            if ''.join(op) == s:
                res.append(op.copy())
                return
            if pos <= len(ip):
                ip1 = ip[pos+1:]
                ip2 = ip
                op1 = op.copy()
                op2 = op.copy()
                op1.append(ip[:pos+1])

                util(0, ip1, op1)
                util(pos+1, ip2, op2)
            return

        util(0, s, [])
        res=list(k for k,_ in groupby(res))

        return res


ob = Solution()
print(ob.wordBreak(5, {"cats", "cat", "and", "sand", "dog"}, 'catsanddog'))

File: DP-CPP/wordBreak.cpp
class Solution
{
public:
    bool f(int ind, int n, string A, unordered_set<string> st)
    {
        if (ind == n)
            return true;
        string temp = "";
        for (int j = ind; j < n; j++)
        {
            temp += A[j];

            if (st.find(temp) != st.end())
            {
                // cout<<' '<<temp;
                if (f(j + 1, n, A, st))
                {

                    return true;
                }
            }
        }
        return false;
    }
    int wordBreak(string A, vector<string> &B)
    {
        // code here

        unordered_set<string> st(B.begin(), B.end());
        int n = A.size();
        return f(0, n, A, st);
    }
};